site/app/libraries/Access.php

```diff
@@ -0,0 +1,248 @@
+    const ALLOW_MIN_FULL_ACCESS_GRADER    = self::ALLOW_INSTRUCTOR | self::ALLOW_FULL_ACCESS_GRADER;
+    const ALLOW_MIN_INSTRUCTOR            = self::ALLOW_INSTRUCTOR;
+
+    /**
+     * @var Core
+     */
+    private $core;
+    private $permissions = [];
+
+    public function __construct(Core $core) {
+        $this->core = $core;
+
+        $this->permissions["grading.status"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP;
+        $this->permissions["grading.status.full"] = self::ALLOW_MIN_FULL_ACCESS_GRADER;
+        $this->permissions["grading.details"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP;
+        $this->permissions["grading.details.show_all"] = self::ALLOW_MIN_FULL_ACCESS_GRADER;
+        $this->permissions["grading.details.show_all_no_sections"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.details.show_empty_teams"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.grade"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT;
+        $this->permissions["grading.grade.if_no_sections_exist"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.save_one_component"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT | self::CHECK_HAS_SUBMISSION;
+        $this->permissions["grading.save_general_comment"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT | self::CHECK_HAS_SUBMISSION;
+        $this->permissions["grading.get_mark_data"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT;
+        $this->permissions["grading.get_gradeable_comment"] = self::ALLOW_MIN_STUDENT | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER | self::CHECK_PEER_ASSIGNMENT_STUDENT;
+        $this->permissions["grading.add_one_new_mark"] = self::ALLOW_MIN_LIMITED_ACCESS_GRADER | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER;
+        $this->permissions["grading.delete_one_mark"] = self::ALLOW_MIN_LIMITED_ACCESS_GRADER | self::CHECK_GRADEABLE_MIN_GROUP | self::CHECK_GRADING_SECTION_GRADER;
+        $this->permissions["grading.get_marked_users"] = self::ALLOW_MIN_LIMITED_ACCESS_GRADER | self::CHECK_GRADEABLE_MIN_GROUP;
+        $this->permissions["grading.get_marked_users.full_stats"] = self::ALLOW_MIN_FULL_ACCESS_GRADER;
+        $this->permissions["grading.show_edit_teams"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.import_teams"] = self::ALLOW_MIN_INSTRUCTOR | self::CHECK_CSRF;
+        $this->permissions["grading.export_teams"] = self::ALLOW_MIN_INSTRUCTOR;
+        $this->permissions["grading.submit_team_form"] = self::ALLOW_MIN_INSTRUCTOR;
```

Comment:
We should condense keys that have the same permission masks unless we absolutely foresee us changing the permission in the very near future (and so why not just do that now?). `save_one_component` and `save_general_comment`, `get_mark_data` and `get_gradeable_comment`, `show_edit_teams` and `export_teams` and `submit_team_form`, `status` and `details`, etc.

Additionally, I foresee a possibility of a time where CSRF token validation would happen automatically which would allow for a bit more consolidation of these.

Reply:
For save/get component vs general, I've added in another check to test for peer graded components, so those aren't combinable anymore. I would combine the team edit permissions except I'm not sure how to handle CSRF.